junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTreeWalker.java

```diff
@@ -57,11 +57,16 @@ private void walk(TestDescriptor globalLockDescriptor, TestDescriptor testDescri
 		}
 		else {
 			Set<ExclusiveResource> allResources = new HashSet<>(exclusiveResources);
-			advisor.forceDescendantExecutionMode(testDescriptor, SAME_THREAD);
-			doForChildrenRecursively(testDescriptor, child -> {
-				allResources.addAll(getExclusiveResources(child));
-				advisor.forceDescendantExecutionMode(child, SAME_THREAD);
-			});
+			if (isReadOnly(allResources)) {
+				doForChildrenRecursively(testDescriptor, child -> allResources.addAll(getExclusiveResources(child)));
+			}
+			if (!isReadOnly(allResources)) {
+				advisor.forceDescendantExecutionMode(testDescriptor, SAME_THREAD);
+				doForChildrenRecursively(testDescriptor, child -> {
+					allResources.addAll(getExclusiveResources(child));
```

Comment:
If we ended up in the first `if` we don't need to collect them here again, do we? Could we rather do what we do in the next `if`, i.e. only force the execution mode? Maybe we should extract that into a method, call it from within a nested `if`, and make this `if` an `else` so we'd end up with sth. like this?

```java
if (isReadOnly(allResources)) {
    doForChildrenRecursively(testDescriptor, child -> allResources.addAll(getExclusiveResources(child)));
    if (!isReadOnly(allResources)) {
        forceDescendantExecutionModeRecursively(advisor, testDescriptor);
    }
} else {
	advisor.forceDescendantExecutionMode(testDescriptor, SAME_THREAD);
	doForChildrenRecursively(testDescriptor, child -> {
		allResources.addAll(getExclusiveResources(child));
		advisor.forceDescendantExecutionMode(child, SAME_THREAD);
	});
}
if (!globalLockDescriptor.equals(testDescriptor) && allResources.contains(GLOBAL_READ_WRITE)) {
    forceDescendantExecutionModeRecursively(advisor, globalLockDescriptor);
    advisor.useResourceLock(globalLockDescriptor, globalReadWriteLock);
}
advisor.useResourceLock(testDescriptor, lockManager.getLockForResources(allResources));
```